lib/ui/dart_runtime_hooks.cc

```diff
@@ -223,4 +226,51 @@ void ScheduleMicrotask(Dart_NativeArguments args) {
   UIDartState::Current()->ScheduleMicrotask(closure);
 }
 
+void LookupClosure(Dart_NativeArguments args) {
+  Dart_Handle closure_name = Dart_GetNativeArgument(args, 0);
+  Dart_Handle library_name = Dart_GetNativeArgument(args, 1);
+  Dart_Handle library;
+  if (library_name == Dart_Null()) {
+    library = Dart_RootLibrary();
+  } else {
+    library = Dart_LookupLibrary(library_name);
+    DART_CHECK_VALID(library);
+  }
+  Dart_Handle closure = Dart_GetClosure(Dart_RootLibrary(), closure_name);
+  Dart_SetReturnValue(args, closure);
+}
+
+void LookupClosureLibrary(Dart_NativeArguments args) {
+  Dart_Handle closure = Dart_GetNativeArgument(args, 0);
+  if (Dart_IsClosure(closure)) {
+    closure = Dart_ClosureFunction(closure);
+    DART_CHECK_VALID(closure);
+  }
+  Dart_Handle owner = closure;
+  Dart_Handle url = Dart_Null();
+  // TOOD(bkonyi): do we need to do this loop?
+  do {
+    owner = Dart_FunctionOwner(owner);
+    if (Dart_IsLibrary(owner)) {
+      url = Dart_LibraryUrl(owner);
+      break;
+    }
+  } while (!Dart_IsLibrary(owner));
```

Comment:
Why do you need this loop, if the owner is not a library object highly likely you are not going to get a library object by looping. This code only works for top level functions. For others you need to get the Class object and then call classobj.library() on it.

Reply:
This loop isn't needed and is already removed in the next patch. Wasn't sure that this worked in the first place (hence the TODO) but I've replaced this code since it didn't work.